### PR TITLE
beam-modules: add fetch-mix-deps

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change
First step in making elixir a first class citizen of Nix
For more informations see https://github.com/NixOS/nixpkgs/issues/105002

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

##### Details

This is a first pass at it, would love to have comments

I've based this off of
https://github.com/hauleth/nix-elixir/blob/master/lib/fetch-mix-deps.nix
and 
https://github.com/cw789/elixir_nix_seed/blob/master/default.nix
Both repos are amazing

##### Decisions that could be controversial

- I've kept the .hex folder in the derivations since it contains the tar files downloaded. The idea on that is to enable reuse of those files. Keeping the .hex in TEMPDIR would destroy those files (as I understand it)

##### Questions that remains

- should we include the version of the program in the dependency derivivation name? I've seen it in some places in the nix ecosystem, not really sure of the for and against on this?
- I'm still not completly sure how we can take advantage of caching so that dependencies for a new version, don't need to be re-downloaded if they were part of a previous version.

#### testing
in order to test you can use this
(assuming the folder link to nixpkgs points to the right location on your local environment)
(the nix_demo repo is basically just a new phoenix project)
```
with import ../../nixpkgs { };

let
  packages = beam.packagesWith beam.interpreters.erlang;
  name = "whatever_floats_your_boat";
  mixEnv = "prod";
  mixSha256 = "0000000000000000000000000000000000000000000000000000";
  version = "0.0.1";
  src = builtins.fetchGit {
    url = "ssh://git@github.com/happysalada/nix_demo";
    rev = "5128e8356d5f58eef495da6ffe40240f3e9324c2";
  };
  mixDeps = packages.fetchMixDeps {
    inherit src name mixEnv version;
    sha256 = mixSha256;
  };

in mixDeps
```

##### Helpful links

[mix docs](https://hexdocs.pm/mix/Mix.html)

### CODE REVIEW

**Comment by happysalada** [comment]
@ankhers When you have a moment, please take a look

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (minijackson) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:23]: Is there a need to export the internal hex directory? Doesn't this make the derivation non-deterministic?

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (minijackson) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:41]: I sometime see fetchers adding `} // removeAttrs args [ "pname" "version", ... ])` at the end so that the fetcher is more "hackable", should that apply here? Do we need to run some postConfigure/Download hooks?

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (minijackson) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:12]: I've had some elixir projects cloning git repository, I think it might be a good idea to add it as a dependency

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (minijackson) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:27]: There is a nice trick from `buildRustCrate`, in which which `fetchCargoTarball` copies the lock file inside `$out` so that the end derivation can [compare it](https://github.com/NixOS/nixpkgs/blob/d884b2d877cfda8247a6ee49f5c26ba3133c3889/pkgs/build-support/rust/default.nix#L164-L190) and verify if the `cargoSha256` is out of date. Maybe we could do that here?

**Review by minijackson — COMMENTED** [review]
Thanks for the PR! This is already looking awesome. Here is some minor things (mostly questions).

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:23]: Since the .hex contains all the tarballs, I was thinking it could be re-used for caching. I have to admit I'm not sure how to do that.
The idea being that if we that gets deleted, then on a new build, we need to re-download all the dependencies again.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:41]: Interested to hear other people's opinion. On my end I can't see use case for making it more hackable. Interesting idea though.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:12]: Thanks! Good catch

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: Very nice find!
Unless I'm wrong, there is no vendored step here that regenerates a mix.lock.
If I understand correctly, mix fetches dependencies from mix.lock. So there couldn't be a difference.
Happy to be proven wrong here.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:23]: So far I haven't found a way to re-use that directory. Unless we find how to re-use that, the default should be to not export it.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (minijackson) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: I think the idea behind that was not that there could be a vendored step that modifies the mix.lock, but that when upgrading a package one often forgets to modify the `mixSha256` (me definitely included). And since it's a fixed-output derivation and it's already in the nix store, nix will happily provide the wrong dependencies.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:23]: I ended up removing the export of the .hex folder.
Here is what I tried.
- giving a .hex folder as an input so that the downloaded files could be used as cache. The problem with that is that if you just give a path name, the output would not be fixed. What I ended up thinking was that the caching step should be taken care of when the process of getting and storing dependency is going through the nix store.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: I think I see what you mean. A `mixSha256` is passed, the content of the downloads should at least be checked against the sha256

Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: I just verified and the sha256 is being checked. As long as null is not supplied then the mixSha256 will be checked. Let me know if I misunderstood your comment

Comment (minijackson) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: Yes, I think we are talking about the same thing. If it is already checked by mix, then I guess there's no need to worry about it ^^

Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:27]: Just to summarize and make sure we are on the same page.
There is only a single mix.lock and mix will get the related packages.
Nix will check that the provided sha256 matches against the dependency that are being fetched.
This can be tested with the instructions on this PR. Upon first fetch, nix will complain about the sha256

**Comment by happysalada** [comment]
@ankhers I think this is ready to merge. Have a look when you have some time.

**Comment by ankhers** [comment]
My apologies. I have been wanting to look at this but I have not found the time. I should be able to look at it over the next couple days.

**Comment by happysalada** [comment]
@ankhers No worries at all, it's wonderful to have your help!

**Comment by happysalada** [comment]
@ankhers just a friendly ping to review this when you have a moment. I understand you are busy, no stress.

**Comment by ankhers** [comment]
Again, I apologize for the delay here.

I guess my only question / concern is that in your example in the original ticket, you are giving `mixSha256` as all zeros. Is that actually allowed, or was that just a mistake in the example?

**Comment by happysalada** [comment]
@ankhers no worries, really greatful for your help here!
I left it with all 0 so that people could confirm that nix would complain and tell them the right mixSha256 to put. I realize now it's not clear. If you prefer, I can update to the real mixSha256. Let me know what works best for you.
(the longer reason I did this, is that if you put null, nix doesn't say anything. So you might think it would work while it doesn't)

**Review by ankhers — COMMENTED** [review]
Everything looks fine. With the caveat that I have not tested it. I can try to test it out as soon as possible. But feel free to merge it if I can't get to it in a timely manner.

**Comment by happysalada** [comment]
@ankhers no worries, take your time. I don't have write access, so I can't merge it.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (yurrriq) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:44]: I could see this potentially being useful for packages with C or other non-BEAM sources.

**Comment by happysalada** [comment]
@ankhers just a friendly ping when you have a moment. No worries if you are busy.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 107440` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by happysalada** [comment]
@SuperSandro2000 If you have a moment, perhaps you can help here? I'm looking to merge this, let me know if you think there is anything. No worries if you are busy of course.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:3]: This confuses me: Are there to inputs?

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:5]: I would prefer if you used lib where necessary and not everywhere.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:10]:
```suggestion
```

Please do not overwrite phases. Disable the ones you do not want to run with ``dontBuild = true;``. downloadPhase is a non standard phase, right? I am not sure if it is allowed in such context.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/fetch-mix-deps.nix:42]:
```suggestion
  impureEnvVars = lib.fetchers.proxyImpureEnvVars;
```

stdenv.lib got recently deprecated and CI should fail when it is used.

@ofborg eval

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:3]: I think you meant 'two' inputs. If you did here is what I think.
This confuses me too. I checked mostly how fetch rebar was implemented.
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/beam-modules/fetch-rebar-deps.nix
The rationale I think is that you want to be able to make several versions of the function.
For example, there is a problem with rebar3 version 3.14 and elixir 1.11
You can make a version fetchMix and a fetchMixRebar3_14 or something.
This is how I understood it. I'm open to suggestions.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:10]: I also took it from fetch rebar deps https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/beam-modules/fetch-rebar-deps.nix
But I thought it was weird as well.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (happysalada) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:44]: I like this feedback. I would say, for the time being, just in the interest of getting this over the line, let's proceed with the current course and we can change this when it is needed.

File: pkgs/development/beam-modules/fetch-mix-deps.nix
Comment (yurrriq) [reply pkgs/development/beam-modules/fetch-mix-deps.nix:3]: The first set of arguments is for the "fetcher" and the second is for the dependency set to be fetched. This is the same pattern used by every other builder and fetcher expression in beam-modules.

**Comment by yurrriq** [comment]
Since I've not yet weighed in on this one... I generally prefer the mix2nix approach over fetchMixDeps, though I haven't followed the evolution of Rust support in Nix lately, and maybe there's something similar there. I'd be curious to compare this with `buildGoModule` too.

**Comment by happysalada** [comment]
Thanks for pitching in, are you talking about this? https://gitlab.com/transumption/mix-to-nix
Just to be sure, fetchMix will only be a step in the buildMix process to ensure you have consistent dependencies.
First fetch the dependencies then build it.
With rust there are many ways to build a rust project, but at least it seems that the process is the same, first fetch cargo deps then build the project.
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/default.nix
Open to suggestions here.

**Comment by yurrriq** [comment]
I meant that in general I would prefer a strategy to convert existing lock files (e.g. `mix.lock`) to Nix expressions, as in node2nix, mvn2nix, etc, rather than adding functions to fetch dependencies. Following along the other discussions though, it seems this is a bit of a nonstarter, at least until the `mix.lock` schema is solidified or some public API is exposed.

So is the idea here to call `fetchMixDeps` and then base the result as `beamDeps` to `buildMix`? We'll want to add some documentation with an example.

**Comment by happysalada** [comment]
That is exactly the idea.

I was thinking of adding the examples and the documentation when I fix buildMix next.

**Comment by yurrriq** [comment]
This seems a happy compromise for now then. I wonder if we should add `fetchDeps ? false` to `buildMix` to streamline the process. Maybe plan to have `buildMix` fetch dependencies by default in 21.03, if we can.

**Comment by happysalada** [comment]
I'm not sure I understand the `fetchDeps ? false` part.
Do you mean that deps would be fetched separately by the user then passed on to buildMix?
I was going for an approach where buildMix would first fetch the dependencies using `fetchMix`. It's true though that in terms of caching it's not ideal, it seems that everytime the source code changes, the deps would be fetched again, I'm not sure how best to proceed there.

**Comment by yurrriq** [comment]
I was thinking the default behaviour of `buildMix` should probably remain the same for now, that is, it expects fetched deps passed as `beamDeps`. It might be nice in the interim, however, to add some way to tell `buildMix` to fetch your dependencies using `fetchMixDeps`, effectively automating the `fetchMixDeps` then pass as `beamDeps` to `buildMix` pattern. Is that any clearer?

**Comment by happysalada** [comment]
Crystal clear, thanks for taking the time to explain.
Let's continue that discussion on the buildMix PR then.
If possible, I don't want to delay this PR further.

**Review by yurrriq — APPROVED** [review]

**Comment by happysalada** [comment]
@Mic92 Just in case you have a bit of time, if you can have a look. In case you are busy, no worries.

**Comment by Mic92** [comment]
> @Mic92 Just in case you have a bit of time, if you can have a look. In case you are busy, no worries.

Looks good. but can you also add a basic usage example to the documentation?
You already provided an example in your PR. There is `doc/languages-frameworks/beam.section.md`.

**Comment by happysalada** [comment]
@Mic92 I was thinking that people would not use this function directly. My idea was that this function would be used inside of the buildMix function. People would then directly use the buildMix function.
I couldn't really figure out a use case for that function on its own.
The test was so that everybody could check that this function worked.
I'm more than happy to add docs if you think it could be useful.

**Comment by happysalada** [comment]
@Mic92 thanks a lot ❤️

**Pull request merged**
